### PR TITLE
Java: Improve AsyncTask data flow support

### PR DESCRIPTION
**Pull request by atorralba**
Uses jump steps to model the life-cycle described here: https://developer.android.com/reference/android/os/AsyncTask\#the-4-steps

### CODE REVIEW

File: java/ql/lib/semmle/code/java/frameworks/android/AsyncTask.qll [hunk 1]
```diff
@@ -4,49 +4,103 @@ import java
...
   ExecuteAsyncTaskMethodAccess() {
-    exists(Method m |
-      this.getMethod() = m and
-      m.getDeclaringType().getSourceDeclaration().getASourceSupertype*() instanceof AsyncTask
-    |
-      m.getName() = "execute" and not m.isStatic() and paramsArgument = this.getArgument(0)
-      or
-      m.getName() = "executeOnExecutor" and paramsArgument = this.getArgument(1)
-    )
+    this.getMethod().hasName(["execute", "executeOnExecutor"]) and
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/frameworks/android/AsyncTask.qll:83]: Member `paramsArgument` is now unbound?

**Review by smowton — COMMENTED** [review]
Looks plausible

File: java/ql/lib/semmle/code/java/frameworks/android/AsyncTask.qll (hunk 1 above)
Comment (atorralba) [reply java/ql/lib/semmle/code/java/frameworks/android/AsyncTask.qll:81]: You're absolutely right, thanks for catching this. Fixed in b75b073.

**Review by smowton — APPROVED** [review]

File: java/ql/lib/semmle/code/java/frameworks/android/AsyncTask.qll [hunk 2]
```diff
@@ -4,49 +4,101 @@ import java
...
  */
 private class AsyncTask extends RefType {
   AsyncTask() { this.hasQualifiedName("android.os", "AsyncTask") }
 }
 
+/** The constructor or initializer method of the `android.os.AsyncTask` class. */
+private class AsyncTaskInit extends Callable {
+  AsyncTaskInit() {
+    this.getDeclaringType().getSourceDeclaration().getASourceSupertype*() instanceof AsyncTask and
+    (this instanceof Constructor or this instanceof InitializerMethod)
```
Comment (intrigus-lgtm) [inline java/ql/lib/semmle/code/java/frameworks/android/AsyncTask.qll:74]: What is `or this instanceof InitializerMethod` doing?
If I remove it, the tests still seem to pass.

Comment (atorralba) [reply java/ql/lib/semmle/code/java/frameworks/android/AsyncTask.qll:74]: Good point, I was missing a test for the initializer method. Thanks for bringing this up! See 03b854a.

**Comment by atorralba** [comment]
@smowton can you re-approve? I only added a test case.

**Review by smowton — APPROVED** [review]

**Pull request merged**